File: 551_600/566. Reshape the Matrix.java
//In MATLAB, there is a very useful function called 'reshape', which can reshape a matrix into a new one with different size but keep its original data.

//You're given a matrix represented by a two-dimensional array, and two positive integers r and c representing the row number and column number of the wanted reshaped matrix, respectively.

//The reshaped matrix need to be filled with all the elements of the original matrix in the same row-traversing order as they were.

//If the 'reshape' operation with given parameters is possible and legal, output the new reshaped matrix; Otherwise, output the original matrix.
public int[][] matrixReshape(int[][] nums,int r,int c){
		if (r*c!=nums.length*nums[0].length) {
			return nums;
		}
		int[][] res = new int[r][c];
		int a =0,b=0;
		for (int i = 0; i < nums.length; i++) {
			for (int j = 0; j < nums[0].length; j++) {
				if (b>=c) {
					a++;
					b = 0;
				}
				res[a][b] = nums[i][j];
				b++;
			}
		}
		return res;
    }

File: 专项练习/Array/566. Reshape the Matrix.cpp
/*
实现reshape函数

Example 1:
Input:
nums =
[[1,2],
 [3,4]]
r = 1, c = 4
Output:
[[1,2,3,4]]
Explanation:
The row-traversing of nums is [1,2,3,4]. The new reshaped matrix is a 1 * 4 matrix, fill it row by row by using the previous list.
Example 2:
Input:
nums =
[[1,2],
 [3,4]]
r = 2, c = 4
Output:
[[1,2],
 [3,4]]
Explanation:
There is no way to reshape a 2 * 2 matrix to a 2 * 4 matrix. So output the original matrix.

 */

class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& nums, int r, int c) {
        if (r * c != nums.size() * nums[0].size()) {
            return nums;
        }
        vector<vector<int>> res(r, vector<int>(c, 0));
        int resI = 0, resJ = 0;
        for (size_t i = 0; i < nums.size(); i++) {
            for (size_t j = 0; j < nums[0].size(); j++) {
                if (resJ < c) {
                    res[resI][resJ] = nums[i][j];
                    resJ++;
                }else{
                    resJ = 0;
                    resI++;
                    res[resI][resJ] = nums[i][j];
                    resJ++;
                }
            }
        }
        return res;
    }
};
